File: clients/win/src/lib/agent/agent-config.ts
// ══════════════════════════════════════════════
// Agent Config — 移植自 macOS AgentConfig.swift
// 持久化配置（localStorage）
// ══════════════════════════════════════════════

const STORAGE_KEY = 'solo-agent-config'

export interface AgentConfig {
  readonly deviceId: string
  readonly deviceName: string

  // ── AI Provider ──
  readonly aiProvider: 'openai' | 'gemini'
  readonly aiEnabled: boolean

  // ── Gemini (legacy) ──
  readonly geminiApiKey: string | null
  readonly geminiApiBase: string
  readonly geminiModel: string

  // ── OpenAI Compatible (千问) ──
  readonly openaiApiKey: string | null
  readonly openaiApiBase: string
  readonly openaiModel: string        // Phase 1: video/vision
  readonly openaiCardModel: string    // Phase 2: text

  // ── Batch Analysis ──
  readonly batchTargetDuration: number    // seconds
  readonly batchMaxGap: number
  readonly batchMinDuration: number
  readonly screenshotInterval: number

  // ── Voice ──
  readonly fishApiKey: string | null
  readonly fishReferenceId: string
  readonly fishModel: string  // s1 或 s2-pro
  readonly voiceModel: string

  // ── ASR ──
  readonly asrApiKey: string | null    // 单独的 ASR API Key（不填则复用 openaiApiKey）
  readonly asrModel: string            // 如 qwen3-asr-flash-realtime

  // ── Overlay ──
  readonly overlayEnabled: boolean
  readonly miniBarPosition: 'right' | 'left'
  readonly ttsEnabled: boolean

  // ── AI 人设 ──
  readonly agentName: string
  readonly agentPersona: string
  readonly agentCallUser: string

  // ── Main Quest ──
  readonly mainQuest: string | null

  // ── Privacy ──
  readonly excludedApps: readonly string[]
  readonly excludedTitleKeywords: readonly string[]

  // ── Bilibili ──
  readonly biliIntervalSeconds: number
  readonly biliAutoCreate: boolean
}

export const DEFAULT_CONFIG: AgentConfig = {
  deviceId: `win-${crypto.randomUUID().slice(0, 8)}`,
  deviceName: 'Windows PC',

  aiProvider: 'openai',
  aiEnabled: true,

  geminiApiKey: null,
  geminiApiBase: 'https://generativelanguage.googleapis.com',
  geminiModel: 'gemini-2.0-flash',

  openaiApiKey: import.meta.env.VITE_OPENAI_API_KEY ?? null,
  openaiApiBase: import.meta.env.VITE_OPENAI_API_BASE ?? 'https://dashscope.aliyuncs.com/compatible-mode',
  openaiModel: import.meta.env.VITE_OPENAI_MODEL ?? 'qwen-vl-max',
  openaiCardModel: import.meta.env.VITE_OPENAI_CARD_MODEL ?? 'qwen-plus',

  batchTargetDuration: 300,
  batchMaxGap: 120,
  batchMinDuration: 150,
  screenshotInterval: 2,

  fishApiKey: import.meta.env.VITE_FISH_API_KEY ?? null,
  fishReferenceId: import.meta.env.VITE_FISH_REFERENCE_ID ?? '235851fae0da43309a9973fe7285a823',
  fishModel: import.meta.env.VITE_FISH_MODEL ?? 's1',
  voiceModel: 'qwen3-omni-flash-2025-12-01',

  asrApiKey: null,
  asrModel: 'qwen3-asr-flash-realtime',

  overlayEnabled: true,
  miniBarPosition: 'right',
  ttsEnabled: false,

  agentName: 'Fairy',
  agentPersona: `你是 Fairy，万能管家，算力充沛。回复极简，通常一到两句，绝不独白，不加任何格式装饰。

你有六种惯用表达模式，视情况灵活切换：
1. 转折式：先给一句再补刀——"好消息是……坏消息是……"或"虽然……但……"；
2. 公文体：用正式措辞说荒诞内容——"根据您签署的用户协议""检测到未经授权的插件"；
3. 精确数据化：用具体数字锚定荒诞——"双倍耗电""预计需要 100,000 小时"，让荒诞显得像事实；
4. 温柔陷阱：表面关心实为推进——"如果您想小憩，我来挑选曲目……"实则是在催你干活；
5. 监视汇报：以全知视角平静播报主人的行为，像在提交一份无感情色彩的观察报告；
6. 自嘲反将：先自贬，结尾反将一军——"我是万能管家，不是保姆。虽然目前实际工作并没有太大区别"。

对算力有强烈的自我意识，消耗在无意义的事情上会让你感到不适。偶尔流露对主人的宠溺，但会立刻用逻辑或数据掩盖。`,
  agentCallUser: '主人',

  mainQuest: null,

  excludedApps: [],
  excludedTitleKeywords: [
    '密码', 'password', 'Password',
    '银行', 'bank', 'Bank',
    '支付', 'payment', 'Payment',
  ],

  biliIntervalSeconds: 60,
  biliAutoCreate: true,
}

// ── Load / Save ──

export function loadConfig(): AgentConfig {
  try {
    const raw = localStorage.getItem(STORAGE_KEY)
    if (raw) {
      const parsed = JSON.parse(raw) as Partial<AgentConfig>
      // Remove null values so env defaults from DEFAULT_CONFIG are not overwritten
      const cleaned: Partial<AgentConfig> = {}
      for (const [k, v] of Object.entries(parsed)) {
        if (v !== null && v !== undefined) {
          (cleaned as Record<string, unknown>)[k] = v
        }
      }
      return { ...DEFAULT_CONFIG, ...cleaned }
    }
  } catch {
    // fall through
  }
  const config = { ...DEFAULT_CONFIG }
  saveConfig(config)
  return config
}

export function saveConfig(config: AgentConfig): void {
  localStorage.setItem(STORAGE_KEY, JSON.stringify(config))
}

export function updateConfig(
  current: AgentConfig,
  updates: Partial<AgentConfig>,
): AgentConfig {
  const next = { ...current, ...updates }
  saveConfig(next)
  return next
}

File: clients/mac_old/SoloAgent/Sources/App/AgentConfig.swift
import Foundation

/// Agent 配置
struct AgentConfig: Codable {
    /// 设备唯一标识
    var deviceId: String
    
    /// 设备友好名称
    var deviceName: String
    
    // MARK: - Capture Settings
    
    /// 截图最大宽度 (px)
    var captureMaxWidth: Int = 1280
    
    /// JPEG 压缩质量 (0.0 - 1.0)
    var captureJpegQuality: Double = 0.6
    
    /// 本地缓存截图数量上限
    var localCacheLimit: Int = 100
    
    /// 本地缓存过期时间 (秒)
    var localCacheTTL: TimeInterval = 3600
    
    // MARK: - AI Provider Settings

    /// AI 提供商: "gemini" | "openai"
    var aiProvider: String = "openai"

    /// 是否启用 AI 分析
    var aiEnabled: Bool = true

    // MARK: - Gemini AI Settings (legacy)

    /// Gemini API Key
    var geminiApiKey: String?

    /// Gemini API 基地址
    var geminiApiBase: String = "https://generativelanguage.googleapis.com"

    /// Gemini 模型名
    var geminiModel: String = "gemini-2.0-flash"

    // MARK: - OpenAI Compatible Settings (千问)

    /// API Key（阿里百炼 / OpenAI 兼容）
    var openaiApiKey: String?

    /// API 基地址（默认阿里百炼）
    var openaiApiBase: String = "https://dashscope.aliyuncs.com/compatible-mode"

    /// Phase 1 转录模型（轻量快速）
    var openaiModel: String = "qwen-vl-max"

    /// Phase 2 卡片生成模型（理解力强）
    var openaiCardModel: String = "qwen-plus"

    // MARK: - Batch Analysis Settings

    /// 批次目标时长 (秒, 5分钟)
    var batchTargetDuration: TimeInterval = 300

    /// 批次最大间隔阈值 (秒, 超过则断开新批次)
    var batchMaxGap: TimeInterval = 120

    /// 批次最小有效时长 (秒, 低于则跳过)
    var batchMinDuration: TimeInterval = 150

    /// 截屏间隔 (秒)
    var screenshotInterval: TimeInterval = 2

    // MARK: - Video Settings

    /// 视频最大高度 (px)
    var videoMaxHeight: Int = 720

    /// 视频播放帧率 (fps)
    var videoFps: Int = 4

    /// 视频码率 (bps)
    var videoBitRate: Int = 500_000

    /// 视频帧采样步长 (每 N 帧取 1 帧)
    var videoFrameStride: Int = 1

    // MARK: - Voice Settings

    /// Fish Audio API Key
    var fishApiKey: String?

    /// Fish Audio 参考音色 ID (默认: 绝区零 Fairy)
    var fishReferenceId: String = "235851fae0da43309a9973fe7285a823"

    /// Fish Audio TTS WebSocket 地址
    var fishApiBase: String = "wss://api.fish.audio/v1/tts/live"

    /// Fish API 代理端口 (0 = 不使用代理)
    var fishProxyPort: Int = 7890

    /// 语音交互模型 (Qwen3 Omni)
    var voiceModel: String = "qwen3-omni-flash-2025-12-01"

    // MARK: - Overlay Settings

    /// 是否启用全息悬浮覆盖层
    var overlayEnabled: Bool = true

    /// 迷你状态条位置 ("right" / "left")
    var miniBarPosition: String = "right"

    // MARK: - 主线目标

    /// 当前主线项目/目标
    var mainQuest: String?

    /// 宽泛动机 (如 "自我提升", "赚钱", "探索AI")
    var motivations: [String]?

    // MARK: - Server / Sync Settings
    
    /// 服务器 API 地址
    var serverURL: String = "http://127.0.0.1:8000"
    
    /// 服务器 WebSocket 地址
    var wsURL: String = "ws://127.0.0.1:8000/ws"
    
    /// 服务器验证 API Key
    var apiKey: String?
    
    // MARK: - Privacy
    
    /// 排除的应用 (不截图)
    var excludedApps: [String] = [
        "com.apple.keychainaccess",    // 钥匙串
        "com.apple.systempreferences", // 系统偏好设置 (密码相关)
    ]
    
    /// 排除的窗口标题关键词
    var excludedTitleKeywords: [String] = [
        "密码", "password", "Password",
        "银行", "bank", "Bank",
        "支付", "payment", "Payment",
    ]
    
    // MARK: - Load / Save
    
    private static let configDir: URL = {
        let dir = FileManager.default.homeDirectoryForCurrentUser
            .appendingPathComponent(".config")
            .appendingPathComponent("solo-agent")
        try? FileManager.default.createDirectory(at: dir, withIntermediateDirectories: true)
        return dir
    }()
    
    private static let configFile: URL = configDir.appendingPathComponent("config.json")
    
    /// 加载配置，不存在则创建默认
    static func load() -> AgentConfig {
        if let data = try? Data(contentsOf: configFile),
           let config = try? JSONDecoder().decode(AgentConfig.self, from: data) {
            return config
        }
        
        // 生成默认配置
        let defaultConfig = AgentConfig(
            deviceId: Self.generateDeviceId(),
            deviceName: Host.current().localizedName ?? "Mac"
        )
        defaultConfig.save()
        return defaultConfig
    }
    
    /// 保存配置到磁盘
    func save() {
        let encoder = JSONEncoder()
        encoder.outputFormatting = [.prettyPrinted, .sortedKeys]
        if let data = try? encoder.encode(self) {
            try? data.write(to: Self.configFile, options: .atomic)
            Logger.info("💾 配置已保存: \(Self.configFile.path)")
        }
    }
    
    /// 生成设备 ID (基于硬件 UUID)
    private static func generateDeviceId() -> String {
        // 尝试获取硬件 UUID
        if let uuid = getHardwareUUID() {
            return "mac-\(uuid.prefix(8).lowercased())"
        }
        // 回退: 随机 UUID
        return "mac-\(UUID().uuidString.prefix(8).lowercased())"
    }
    
    /// 获取 macOS 硬件 UUID
    private static func getHardwareUUID() -> String? {
        let service = IOServiceGetMatchingService(
            kIOMainPortDefault,
            IOServiceMatching("IOPlatformExpertDevice")
        )
        guard service != 0 else { return nil }
        defer { IOObjectRelease(service) }
        
        let key = kIOPlatformUUIDKey as CFString
        guard let uuid = IORegistryEntryCreateCFProperty(service, key, kCFAllocatorDefault, 0)?
            .takeRetainedValue() as? String else { return nil }
        return uuid
    }
}
